File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/ToolRegistry.fs
namespace Skight.AgentPlatform.FSharp

open System.Collections.Generic
open System.Text.Json

type ToolRegistry() =
    let tools = Dictionary<string, string -> Async<string>>()
    let schemas = List<ToolSchema>()

    member this.Register(name: string, description: string, handler: string -> Async<string>, parametersJson: string) =
        tools.[name] <- handler

        use _ = JsonDocument.Parse(parametersJson)
        let schema = {
            Name = name
            Description = description
            ParametersJson = parametersJson
        }
        schemas.Add(schema)

    member this.ExecuteToolAsync(name: string, argumentsJson: string) : Async<string> =
        async {
            match tools.TryGetValue(name) with
            | true, handler ->
                try
                    return! handler argumentsJson
                with ex ->
                    return sprintf "Error executing tool '%s': %s" name ex.Message
            | false, _ ->
                return sprintf "Error: Tool '%s' not found." name
        }

    member this.AsExecutor : ToolExecutor =
        fun name args -> this.ExecuteToolAsync(name, args)

    member this.GetToolSchemas() : ToolSchema list =
        Seq.toList schemas

    member this.GetRegisteredNames() : string list =
        Seq.toList tools.Keys

File: implementations/csharp/src/Skight.AgentPlatform/ToolRegistry.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Azure.AI.OpenAI;

namespace Skight.AgentPlatform
{
    public class ToolRegistry
    {
        private readonly Dictionary<string, Func<string, Task<string>>> _tools = new();
        private readonly List<FunctionDefinition> _schemas = new();

        public void Register(string name, string description, Func<string, Task<string>> func, string parametersJson)
        {
            _tools[name] = func;
            
            using var doc = JsonDocument.Parse(parametersJson);
            var functionDef = new FunctionDefinition
            {
                Name = name,
                Description = description,
                Parameters = BinaryData.FromString(parametersJson)
            };
            _schemas.Add(functionDef);
        }

        public async Task<string> ExecuteToolAsync(string name, string argumentsJson)
        {
            if (!_tools.TryGetValue(name, out var func))
            {
                return $"Error: Tool '{name}' not found.";
            }

            try
            {
                return await func(argumentsJson);
            }
            catch (Exception ex)
            {
                return $"Error executing tool '{name}': {ex.Message}";
            }
        }

        public List<FunctionDefinition> GetToolSchemas()
        {
            return _schemas;
        }
    }
}
